File: legacy/UserInterface/uart.py
import serial

class DataTypes:
    DOMAIN_DATA_TYPE = 0
    USERNAME_DATA_TYPE = 1
    PASSWORD_DATA_TYPE = 2

###################################################################
#
#   get_pw_count()
#   Description:
#   Writes the read password count command and receives every byte
#   to get up to date number of bytes. This is typically done at
#   the start so that the python program can get an active count 
#   to keep track of and update
#
###################################################################
def get_pw_count():
    ser = serial.Serial("/dev/ttyUSB1", baudrate=115200)
    ser.write(bytes.fromhex("FF")) 
    ser.write(bytes.fromhex("00"))
    readstr = b""

    for i in range(64):
        readstr = ser.read(1) + readstr # Receive the count

    ser.close()
    return int.from_bytes(readstr, byteorder='big') # Return big endian number

###################################################################
#
#   set_pw_count(num_pw)
#   Parameter:
#   - num_pw (int): Number of passwords to write  
#   Description:
#   Writes the write password count command and gives new password
#   count
#
###################################################################
def set_pw_count(num_pw: int):
    ser = serial.Serial("/dev/ttyUSB1", baudrate=115200)
    ser.write(bytes.fromhex("7F")) # Read password count command
    ser.write(bytes.fromhex("00")) # Doesn't matter what we write here

    for i in range(64-len(int.to_bytes(num_pw))):
        ser.write(bytes.fromhex('00')) # Pad front with zeros to fill up expected 64 bytes
    ser.write(int.to_bytes(num_pw)) # Write actual data
    
    readstr = ""
    for i in range(4):
        readstr = ser.read(1).decode("ascii") + readstr # Read status (Should just be "done")

    ser.close()
    return readstr

##################################################################
#
#   get_data(cmd, address)
#   parameters:
#   - cmd (int): Command saying to perform a read and what data 
#   type to retreive
#   - address (int): Address to retrieve data from
#   Description:
#   Function takes in string of command to convert to byte 
#   string, and the address as an integer and performs a read
#
##################################################################
def get_data(cmd: str, address: int):
    ser = serial.Serial("/dev/ttyUSB1", baudrate=115200) # Create serial session
    ser.write(bytes.fromhex(cmd)) # Write command
    ser.write(int.to_bytes(address)) # Write the address
    readstr = ""

    for i in range(64):
        readstr = ser.read(1).decode("ascii") + readstr # Read all 64 bytes

    ser.close()
    return readstr

###################################################################
#
#   set_data(cmd, address, data)
#   parameters:
#   - cmd (int): Command saying to perform a write and what data 
#   type to write
#   - address (int): Address to write data to
#   - data (str): Data to write
#   Description:
#   Function takes in string of command to convert to byte 
#   string, and the address as an integer and performs a write
#
###################################################################
def set_data(cmd: str, address: int, data: str):
    ser = serial.Serial("/dev/ttyUSB1", baudrate=115200)
    ser.write(bytes.fromhex(cmd)) # Send command
    ser.write(int.to_bytes(address)) # Send address
    
    for i in range(64-len(data)):
        ser.write(bytes.fromhex('00')) # Pads front with zeros
    ser.write(data) # Write the actual data
    
    readstr = ""
    for i in range(4):
        readstr = ser.read(1).decode("ascii") + readstr # Read status (Should just be "done")

    ser.close()
    return readstr

###################################################################
#
#   read(data_type, address)
#   parameters:
#   - data_type (int): What type of data to update
#   - address (int): Address to read data from
#   Description:
#   Takes in constant int type of data and calls the get_data 
#   function.
#
###################################################################
def read(data_type: int, address: int):
    cmd = 80
    match data_type:
        case DataTypes.DOMAIN_DATA_TYPE:
            cmd += 0 # Start at offset 0
        case DataTypes.USERNAME_DATA_TYPE:
            cmd += 2 # Read from offset 2
        case DataTypes.PASSWORD_DATA_TYPE:
            cmd += 3 # Read from offset 3
        case _:
            print("Not a valid data type")
            return ""
    
    readstr = get_data(str(cmd), address)
    if(data_type == 0):
        readstr += get_data(str(cmd+1), address) # Read from the second 64 bytes of the domain names
    return readstr

###################################################################
#
#   write(data_type, address, data)
#   parameters:
#   - data_type (int): What type of data to update 
#   type to write
#   - address (int): Address to write data to
#   - data (str): Data to write
#   Description:
#   Takes in constant int type of data to write to and calls the 
#   set_data function.
#
###################################################################
def write(data_type: int, address: int, data: str):
    cmd = 0
    match data_type:
        case DataTypes.DOMAIN_DATA_TYPE:
            cmd += 0 # Start at offset 0
        case DataTypes.USERNAME_DATA_TYPE:
            cmd += 2 # Write to offset 2
        case DataTypes.PASSWORD_DATA_TYPE:
            cmd += 3 # Write to offset 3
        case _:
            print("Not a valid data type")
            return ""
    
    readstr = set_data("0" + str(cmd), address, data[0:63].encode("ascii")) # Want to substring the data up to 64 characters (64 bytes)
    if(data_type == 0):
        readstr = set_data("0" + str(cmd+1), address, data[63:].encode("ascii")) # Write to second 64 bytes of the domain names
    return readstr

###################################################################
#
#   listDomains(num_pw)
#   parameters:
#   - num_pw (int): Password count
#   Description:
#   Lists each domain name
#
###################################################################
def listDomains(num_pw: int):
    print("----------Domains----------")
    for i in range(num_pw):
        readstr = read(DataTypes.DOMAIN_DATA_TYPE, i) # Read domain name of each iteration of the loop
        print(f"{i+1}. {readstr}")
    print("---------------------------")

###################################################################
#
#   listDomainInfo(num_pw)
#   parameters:
#   - num_pw (int): Password count
#   - domain (int): Domain selection
#   Description:
#   Reads all piece of data of given 
#
###################################################################
def listDomainInfo(domain: int, num_pw: int):
    if(domain > num_pw): # Return if given invalid domain
        print("This is not a valid domain")
        return

    # Read and print out each piece of data for requested domain
    readstr = read(0, domain-1)
    print("Domain:", readstr)
    readstr = read(1, domain-1)
    print("Username:", readstr)
    readstr = read(2, domain-1)
    print("Password:", readstr)

###################################################################
#
#   createPassword(domain, username, password, num_pw)
#   parameters:
#   - domain (str): New domain name
#   - username (str): New username
#   - password (str): New password
#   - num_pw (int): Password count
#   Description:
#   Writes new domain name, username, and password, as well as
#   increments the password count
#
###################################################################
def createPassword(domain: str, username: str, password: str, num_pw: int):
    # Write to the FPGA
    write(DataTypes.DOMAIN_DATA_TYPE, num_pw, domain)
    write(DataTypes.USERNAME_DATA_TYPE, num_pw, username)
    write(DataTypes.PASSWORD_DATA_TYPE, num_pw, password)

    # Increments password count and updates it
    num_pw += 1
    readstr = set_pw_count(num_pw)
    return num_pw

###################################################################
#
#   changeUsername(domain, username, num_pw)
#   parameters:
#   - domain (int): Domain Selection
#   - username (str): New Username
#   - num_pw (int): Password count
#   Description:
#   Change username of given domain
#
###################################################################
def changeUsername(domain: int, username: str, num_pw: int):
    if(domain > num_pw): # Trying to change username of a domain name that doesn't exist
        print("This is not a valid domain")
        return
    else:
        write(DataTypes.USERNAME_DATA_TYPE, domain-1, username) # Write new username

###################################################################
#
#   changePassword(domain, password, num_pw)
#   parameters:
#   - domain (int): Domain Selection
#   - password (str): New password
#   - num_pw (int): Password count
#   Description:
#   Change password of given domain
#
###################################################################
def changePassword(domain: int, password: str, num_pw: int):
    if(domain > num_pw): # Trying to change password of a domain name that doesn't exist
        print("This is not a valid domain")
        return
    else:
        write(DataTypes.PASSWORD_DATA_TYPE, domain-1, password) # Write new password

###################################################################
#
#   deleteDomain(domain, num_pw)
#   parameters:
#   - domain (int): Domain Selection
#   - num_pw (int): Password count
#   Description:
#   Starting from selected domain, reads the next domain, username,
#   and password and replacing the current domain. Zeroizes the last
#   domain and decrements the password count.
#
####################################################################
def deleteDomain(domain: int, num_pw: int):
    # Want to make sure the user is sure they want to absolutely remove the domain
    confirm = input("Are you sure? (Deleting this will zeroize out the data and can only be retreived manually) (y/N): ")
    if(confirm == "y"):
        pass
    else:
        return num_pw
    
    # Replace each domain to next one
    curr_pw = domain
    while(curr_pw < num_pw):
        new_domain = read(0, curr_pw)
        write(0, curr_pw-1, new_domain[1:128])
        new_username = read(1, curr_pw)
        write(1, curr_pw-1, new_username[1:64])
        new_password = read(2, curr_pw)
        write(2, curr_pw-1, new_password[1:64])
        curr_pw += 1
    
    # Zeroize last entry
    write(DataTypes.DOMAIN_DATA_TYPE, curr_pw-1, "")
    write(DataTypes.USERNAME_DATA_TYPE, curr_pw-1, "")
    write(DataTypes.PASSWORD_DATA_TYPE, curr_pw-1, "")
    
    # Decrement and update password count
    set_pw_count(num_pw - 1)
    return num_pw - 1

###################################################################
#
#   int()
#   Description:
#   Gets the stored password count, displays it and all domains
#
###################################################################
def init():
    num_pw = get_pw_count()
    print("There are", num_pw, "password(s) stored.")
    listDomains(num_pw)
    return num_pw

# Example UI for the functions above, to be replaced by GUI
def run():
    choice = 0
    num_pw = init()

    while(choice != 7):
        print("\nWhat would you like to do?")
        print("1. List Domains")
        print("2. List Domain Information")
        print("3. Add Password")
        print("4. Change Username")
        print("5. Change Password")
        print("6. Remove Password")
        print("7. Quit")
        choice = int(input("What would you like to do?: "))

        if(choice == 1):
            listDomains(num_pw)
        elif(choice == 2):
            listDomains(num_pw)
            domain = int(input("Which domain would you like to look at?: "))
            listDomainInfo(domain, num_pw)
        elif(choice == 3):
            domain = input("What is the domain name: ")
            username = input("What is your username: ")
            password = input("What is your password: ")
            num_pw = createPassword(domain, username, password, num_pw)
        elif(choice == 4):
            listDomains(num_pw)
            domain = int(input("Which domain would you like to change the username of: "))
            username = input("What is the new username: ")
            changeUsername(domain, username, num_pw)
        elif(choice == 5):
            listDomains(num_pw)
            domain = int(input("Which domain would you like to change the password of: "))
            password = input("What is the new password: ")
            changePassword(domain, password, num_pw)
        elif(choice == 6):
            listDomains(num_pw)
            domain = int(input("Which domain would you like to remove: "))
            num_pw = deleteDomain(domain, num_pw)
        elif(choice == 7):
            break

if __name__ == "__main__":
    run()

File: keylimepi_ui/src/uart.rs
use serialport::available_ports;
use std::time::Duration;

const DOMAIN_CODE: u32 = 0;
const USERNAME_CODE: u32 = 1;
const PASSWORD_CODE: u32 = 2;

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// # Description:
///   Writes the read password count command and receives every byte
///   to get up to date number of bytes. This is typically done at
///   the start so that the python program can get an active count 
///   to keep track of and update
fn get_pw_count(path: &String) -> u32 {
    // Open the port
    let mut port = serialport::new(path, 115_200)
        .timeout(Duration::from_millis(100))
        .open()
        .expect("Failed to open port");

    // Send the "read pw count" command and address 0 as filler
    let cmd: [u8; 2] = [255, 0];
    port.write(&cmd).expect("Get Password command failed");

    // Read 64 bytes from the FPGA
    let mut serial_buf: Vec<u8> = vec![0; 64];
    port.read(serial_buf.as_mut_slice()).expect("Failed to receive number of passwords");

    // concatenate each byte into one 32 bit integer
    let num_pw = ((serial_buf[3] as u32) << 24) | 
                        ((serial_buf[2] as u32) << 16) | 
                        ((serial_buf[1] as u32) << 8) | 
                        (serial_buf[0] as u32);

    return num_pw;
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords to write
/// # Description:
///   Writes the write password count command and gives new password
///   count
fn set_pw_count(path: &String, num_pw: u32) -> String {
    // Open port
    let mut port = serialport::new(path, 115_200)
        .timeout(Duration::from_millis(100))
        .open()
        .expect("Failed to open port");

    // Send the "read pw count" command then send address 0 as filler
    let cmd: [u8; 2] = [127, 0];
    port.write(&cmd).expect("Get Password command failed");

    // Create empty array of 64 bytes, and fill with little endian representation of integer
    let mut output: [u8; 64] = [0; 64];
    let num_output: [u8; 4] = num_pw.to_le_bytes();
    for i in 0..4 {
        output[63-i] = num_output[i];
    }

    // Write final full 64 byte array
    port.write(&output).expect("Write failed!");

    // Receive "Done" status and return
    let mut serial_buf: Vec<u8> = vec![0; 4];
    port.read(serial_buf.as_mut_slice()).expect("Failed to get status for setting data");

    let write_stat = match str::from_utf8(&serial_buf) {
        Ok(v) => v,
        Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
    };
    
    // Need to reverse the string (TODO: Should look into reversing the data on FPGA)
    return write_stat.to_string().chars().rev().collect::<String>();
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - cmd (u8): The command to send to receive data.
/// - address (u8): Address to read from
/// # Description:
/// Function takes in string of command to convert to byte 
/// array, and the address as an integer and performs a read
fn get_data(path: &String, cmd: u8, address: u8) -> String {
    // Open port
    let mut port = serialport::new(path, 115_200)
        .timeout(Duration::from_millis(100))
        .open()
        .expect("Failed to open port");

    // Send the command and address
    let setup: [u8; 2] = [cmd, address];
    port.write(&setup).expect("Get Data command failed");

    // Read the full 64 bytes
    let mut serial_buf: Vec<u8> = vec![0; 64];
    port.read(serial_buf.as_mut_slice()).expect("Failed to retreive data");

    // Convert vector to string
    let data = match str::from_utf8(&serial_buf) {
        Ok(v) => v,
        Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
    };

    // Return string reversed
    return data.to_string().chars().rev().collect::<String>();
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - cmd (u8): The command to send to transmit data.
/// - address (u8): Address to write to
/// - data (String): 
/// # Description:
/// Function takes in string of command to convert to byte 
/// array, and the address as an integer and performs a write
fn set_data(path: &String, cmd: u8, address: u8, data: String) -> String {
    // Open port
    let mut port = serialport::new(path, 115_200)
        .timeout(Duration::from_millis(100))
        .open()
        .expect("Failed to open port");

    // Send the command and address
    let setup: [u8; 2] = [cmd, address];
    port.write(&setup).expect("Set Data command failed");

    // Write to the full 64 bytes, filling in the rest with 0
    let mut output: [u8; 64] = [0; 64];
    for i in 0..data.len() {
        output[i] = data.as_bytes()[i];
    }
    port.write(&output).expect("Failed to set data");

    // Get response word (only expecting "Done")
    let mut serial_buf: Vec<u8> = vec![0; 4];
    port.read(serial_buf.as_mut_slice()).expect("Failed to get status for setting data");

    // Convert to String
    let write_stat = match str::from_utf8(&serial_buf) {
        Ok(v) => v,
        Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
    };

    // Return reversed
    return write_stat.to_string().chars().rev().collect::<String>();
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - data_type (u32): The type of data to read from
/// - address (u8): Address of page to read from
/// # Description:
/// Takes in constant int type of data and calls the get_data 
/// function.
fn read(path: &String, data_type: u32, address: u8) -> String {
    // Set command based on the data type
    let cmd: u8 = match data_type {
        DOMAIN_CODE => 128, // 0x80
        USERNAME_CODE => 130, // 0x82
        PASSWORD_CODE => 131, // 0x83
        _ => 128,
    };

    // Get the data based on the command and address
    let data: String = get_data(path, cmd, address);

    // If the data type is domain, we need to read the second half of it
    if data_type == DOMAIN_CODE {
        let second_half: String = get_data(path, cmd+1, address);

        let data = format!("{second_half}{data}");

        return data;
    }

    return data;
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - data_type (u32): The type of data to write
/// - address (u8): Address of page to write to
/// - data (String): Data to write
/// # Description:
/// Takes in constant int type of data to write to and calls the 
/// set_data function.
fn write(path: &String, data_type: u32, address: u8, data: String) -> String {
    // Set command based on data type
    let cmd: u8 = match data_type {
        DOMAIN_CODE => 0, // 0x00
        USERNAME_CODE => 2, // 0x02
        PASSWORD_CODE => 3, // 0x03
        _ => 0, // 0x00
    };

    // If the data type is domain, we need to make two writes
    if data_type == DOMAIN_CODE {
        let first_half: String;
        let second_half: String;

        // The 128 bytes need to be padded
        // If the string is larger than 64 bytes, it needs to be split
        if data.len() > 64 {
            first_half = data[0..63].trim().to_string();
            second_half = data[64..].to_string();   
        } else {
            first_half = data.trim().to_string();
            second_half = String::from("");
        }

        // Send the two writes and get the status of the second write because
        // if the first one doesn't work, the second one wouldn't either
        set_data(path, cmd, address, first_half);
        let write_stat: String = set_data(path, cmd+1, address, second_half);

        return write_stat;
    } else {
        // Just need to send the data once 
        let write_stat: String = set_data(path, cmd, address, data.trim().to_string());

        return write_stat;
    }
}

/// # Description:
/// This function is meant to be called at the start and returns
/// the path of the port and the number of passwords. Loops through
/// every possible path to open up a port. If opening up a port, 
/// does not error, get the number of passwords.
pub fn find_path() -> (String, u32) {
    // Initialize the data and get all available ports
    let mut num_pw: u32 = 0;
    let ports = available_ports().expect("No ports found!");
    
    // Loop through every serialport
    for p in ports {
        // If we're able to connect to port, attempt to get number of passwords
        let port_name: String = p.port_name.clone();
        if let Ok(port) = serialport::new(p.port_name, 115_200).timeout(Duration::from_millis(100)).open() {
            drop(port);
            let path: String = port_name;
            num_pw = get_pw_count(&path);
            return (path, num_pw);
        }
        
    }

    // If none matches, return "None"
    let path: String = String::from("None");

    return (path, num_pw);
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// # Description:
/// Returns a vector of all domains
pub fn list_domains(path: &String, num_pw: u32) -> Vec<String>{
    // Initialize domains vector
    let mut domains: Vec<String> = Vec::new();

    // Loop through number of passwords
    // Read the domain at i, then push to the vector
    for i in 0..num_pw {
        let domain: String = read(path, DOMAIN_CODE, i as u8);
        domains.push(domain);
    }

    return domains;
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// - domain (u8): Choice of domain
/// # Description:
/// Returns string of domain, username, and password
pub fn list_domain_info(path: &String, num_pw: u32, domain: u8) -> (String, String, String) {
    // If domain is greater than number of passwords, returns error as stirng
    if domain > num_pw as u8 {
        println!("This is not a valid domain");
        return ("Error".to_string(), "Error".to_string(), "Error".to_string());
    }

    // Performs reads of each data type, then returns them
    let domain_str: String = read(path, DOMAIN_CODE, domain-1);
    let username_str: String = read(path, USERNAME_CODE, domain-1);
    let password_str: String = read(path, PASSWORD_CODE, domain-1);

    return (domain_str, username_str, password_str);
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// - domain (String): String of new domain
/// - username (String): String of new username
/// - password (String): String of new password
/// # Description:
/// Adds new password entry
pub fn create_domain(path: &String, num_pw: u32, domain: String, username: String, password: String) -> (u32, String) {
    // Write each data type to number of passwords address (Writing to new page)
    write(path, DOMAIN_CODE, num_pw as u8, domain);
    write(path, USERNAME_CODE, num_pw as u8, username);
    write(path, PASSWORD_CODE, num_pw as u8, password);

    // Increment number of passwords and update the keylimepi
    let new_num_pw = num_pw + 1;
    let write_stat: String = set_pw_count(path, new_num_pw);
    return (new_num_pw, write_stat);
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// - domain (u8): Choice of domain
/// - username (String): String of new username
/// # Description:
/// Updates the username field in the given domain entry
pub fn change_username(path: &String, num_pw: u32, domain: u8, username: String) -> String{
    let write_stat: String;
    if domain > num_pw as u8 {
        println!("This is not a valid domain");
        write_stat = "InvalidAddr".to_string();
    } else {
        write_stat = write(path,USERNAME_CODE, domain-1, username);
    }

    return write_stat;
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// - domain (u8): Choice of domain
/// - password (String): String of new password
/// # Description:
/// Updates the password field in the given domain entry
pub fn change_password(path: &String, num_pw: u32, domain: u8, password: String) -> String {
    let write_stat: String;
    if domain > num_pw as u8 {
        println!("This is not a valid domain");
        write_stat = "InvalidAddr".to_string();
    } else {
        write_stat = write(path,PASSWORD_CODE, domain-1, password);
    }

    return write_stat;
}

/// # Parameters:
/// - path (&String ): Path to device to open the serial port
/// - num_pw (u32): Number of passwords
/// - domain (u8): Choice of domain
/// # Description:
/// Starting from selected domain, reads the next domain, username,
/// and password and replacing the current domain. Zeroizes the last
/// domain and decrements the password count.
pub fn delete_domain(path: &String, num_pw: u32, domain: u8) -> (u32, String) {
    // Start at given domain and loop through each entry
    let mut curr_pw: u8 = domain;
    while curr_pw < num_pw as u8 {
        // Write the next entry into current entry
        let new_domain: String = read(path, DOMAIN_CODE, curr_pw);
        write(path, DOMAIN_CODE, curr_pw-1, new_domain);
        let new_username: String = read(path, USERNAME_CODE, curr_pw);
        write(path, USERNAME_CODE, curr_pw-1, new_username);
        let new_password: String = read(path, PASSWORD_CODE, curr_pw);
        write(path, PASSWORD_CODE, curr_pw-1, new_password);
        curr_pw += 1;
    }

    // Write all 0 into the last entry to zeroize it
    write(path, DOMAIN_CODE, curr_pw-1, "".to_string());
    write(path, USERNAME_CODE, curr_pw-1, "".to_string());
    write(path, PASSWORD_CODE, curr_pw-1, "".to_string());

    // Decrement number of passwords and update entry on device
    let new_num_pw: u32 = num_pw - 1;
    let write_stat: String = set_pw_count(path, new_num_pw);
    return (new_num_pw, write_stat);
}
